### PR TITLE
fix: [CDS-48389]: add support to datetime types when passed by workflow variable on jira custom field

### PR DESCRIPTION
**Pull request by rafaelmart1ns**
## Describe your changes
When the user passes the jira custom field datetime values as a workflow variable, the error reported by the client occurs. The reason is that the regex handled in the backend expect a pre-processing done by the UI.
Below are the examples being treated in pr:
`yyyy/MM/dd HH:mm` - Format of UI
`yyyy-MM-ddTHH:mm:ss.Z;+-HH:mm`
`current() +2h 2d` or any variation that meets the documentation.
 
how the UI pre-processes the data before sending this PR to the backend creates support for the standards without any pre-processing.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)

### CODE REVIEW

**Comment by rafaelmart1ns** [comment]
trigger ft1

**Comment by rafaelmart1ns** [comment]
trigger ft1

File: 400-rest/src/main/java/software/wings/sm/states/collaboration/JiraCreateUpdate.java [hunk 1]
```diff
@@ -125,6 +126,17 @@ public class JiraCreateUpdate extends State implements SweepingOutputStateMixin
...
   private static final String NUMBER = "number";
 
+  private static final HashMap<String, Long> MILLISECONDS_PER_UNIT = new HashMap<>() {
+    {
+      put("ms", 1L);
+      put("s", 1_000L);
+      put("m", 60_000L);
+      put("h", 3_600_000L);
+      put("d", 86_400_000L);
+      put("w", 604_800_000L);
```
Comment (lasdox) [inline 400-rest/src/main/java/software/wings/sm/states/collaboration/JiraCreateUpdate.java:136]: Use Duration lib here for better readability

File: 400-rest/src/test/java/software/wings/sm/states/collaboration/JiraCreateUpdateTest.java [hunk 2]
```diff
@@ -231,6 +232,63 @@ public void testParseDateTimeValue() {
...
+    expectedVal = String.valueOf(System.currentTimeMillis() - 172_800_000L);
+    assertThat(parsedVal).isEqualTo(expectedVal);
+
+    fieldValue = "current() *2d";
+    String finalFieldValue = fieldValue;
+    assertThatThrownBy(() -> jiraCreateUpdateState.parseDateTimeValue(finalFieldValue, context))
+        .isInstanceOf(InvalidRequestException.class)
+        .hasMessage("Cannot parse date time value from current() *2d");
+
+    fieldValue = "2022/12/24 10:00";
```
Comment (lasdox) [inline 400-rest/src/test/java/software/wings/sm/states/collaboration/JiraCreateUpdateTest.java:286]: this pattern should fail, we need timezone

Comment (rafaelmart1ns) [reply 400-rest/src/test/java/software/wings/sm/states/collaboration/JiraCreateUpdateTest.java:286]: I will add support to the timezone too.

File: 400-rest/src/main/java/software/wings/sm/states/collaboration/JiraCreateUpdate.java [hunk 3]
```diff
@@ -151,6 +163,11 @@ public class JiraCreateUpdate extends State implements SweepingOutputStateMixin
   private Map<String, JiraCustomFieldValue> customFields;
   private static final Pattern currentPattern =
       Pattern.compile("(current\\(\\))(\\s*([+-])\\s*(\\d{0,13}))*", Pattern.CASE_INSENSITIVE);
+
+  private static final Pattern uiPattern =
+      Pattern.compile("^\\d{4}\\/\\d{2}\\/\\d{2}\\s*\\d{2}:\\d{2}$", Pattern.CASE_INSENSITIVE);
```
[the remark this answers is not in the document]
Comment (rafaelmart1ns) [reply 400-rest/src/main/java/software/wings/sm/states/collaboration/JiraCreateUpdate.java:168]: In this case, as the same case when using the UI calendar. It should use UTC in the case of usage of this pattern.

**Review by lasdox — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)* status: 3 file(s) to review. 3. unresolved discussions. 1 of 2 approvals obtained.

**Review by Fernando-Dourado — APPROVED** [review]
> Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)* status: 3. unresolved discussions.

**Review by Fernando-Dourado — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)* status: 3. unresolved discussions.

**Review by Fernando-Dourado — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899#-:-NJzhO4ZEi6EJ_lvWi5Y:bnfp4nl)* status: 3. unresolved discussions.

**Review by rafaelmart1ns — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 2 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)* status: 3. unresolved discussions.

**Review by lasdox — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41899)* status: :shipit: complete! ✓

**Pull request merged**